paywall/src/windowTypes.ts

```diff
@@ -0,0 +1,123 @@
+import { PaywallConfig } from './unlockTypes'
+
+// window sub-types
+
+export interface EventDocument {
+  createEvent: (type: string) => CustomEvent
+}
+
+export interface IframeManagingDocument {
+  createElement: (type: 'iframe') => IframeType
+  querySelector: (selector: string) => any
+  body: {
+    insertAdjacentElement: (where: 'afterbegin', iframe: IframeType) => void
+    style: {
+      overflow: string
+    }
+  }
+}
+
+export interface FullDocument extends EventDocument, IframeManagingDocument {}
+
+// used in unlock.js/dispatchEvent.ts
+export interface EventWindow {
+  CustomEvent: {
+    // this is copy/paste from the .d.ts for window, because it
+    // declares CustomEvent as an interface and exports it, but
+    // also as a global variable, and the global variable is not
+    // also exported in the definition of window.
+    // So, calling new window.CustomEvent() fails, but
+    // new CustomEvent() succeeds.
+    prototype: CustomEvent
+    new <T>(typeArg: string, eventInitDict?: CustomEventInit<T>): CustomEvent<T>
+  }
+  document: EventDocument
+  dispatchEvent: (event: CustomEvent) => void
+}
+
+// used anywhere cache is used
+export interface LocalStorageWindow {
+  localStorage: Storage
+}
+
+// used in web3Proxy.ts
+export interface web3MethodCall {
+  method: string
+  params: any[]
+  jsonrpc: '2.0'
+  id: number
+}
+
+export type web3Callback = (error: Error | string | null, result: any) => void
+export type web3Send = (
+  methodCall: web3MethodCall,
+  callback: web3Callback
+) => void
+
+export interface Web3Window extends PostOfficeWindow {
+  web3?: {
+    currentProvider: {
+      sendAsync?: web3Send
+      send?: web3Send
+      isMetamask: true | undefined
+    }
+  }
+}
+
+// used in utils/postOffice.ts
+export interface MessageEvent {
+  source: any
+  origin: string
+  data: any
+}
+
+export type MessageHandler = (event: MessageEvent) => void
+
+export interface PostOfficeWindow {
+  addEventListener: (type: 'message', handler: MessageHandler) => void
```

Comment:
if `type` is a fixed value, I think we  should reduce the parameters of this method

Reply:
again, the signature is from the `window` object, this is just restricting the signature further for our source, which should only register a listener for `'message'` event type